File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/jmm/JmmNMM.kt
package info.bagen.rust.plaoc.microService.sys.jmm

import info.bagen.rust.plaoc.datastore.JmmMetadataDB
import info.bagen.rust.plaoc.microService.core.BootstrapContext
import info.bagen.rust.plaoc.microService.core.NativeMicroModule
import info.bagen.rust.plaoc.microService.helper.*
import info.bagen.rust.plaoc.microService.sys.dns.nativeFetch
import info.bagen.rust.plaoc.microService.sys.jmm.ui.JmmManagerActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import org.http4k.core.Method
import org.http4k.core.Uri
import org.http4k.core.query
import org.http4k.lens.Query
import org.http4k.lens.string
import org.http4k.routing.bind
import org.http4k.routing.routes

class JmmNMM : NativeMicroModule("jmm.sys.dweb") {
    companion object {
        private val apps = mutableMapOf<Mmid, JsMicroModule>()

        fun getAndUpdateJmmNmmApps() = apps

        suspend fun NativeMicroModule.nativeFetchFromJS(mmid: Mmid) {
            nativeFetch(
                Uri.of("file://dns.sys.dweb/open")
                    .query("app_id", mmid.encodeURIComponent())
            )
        }

        suspend fun NativeMicroModule.nativeFetchInstallDNS(jmmMetadata: JmmMetadata) { // 安装完成后，需要注册到 DnsNMM 中
            nativeFetch(
                Uri.of("file://dns.sys.dweb/install")
                    .query("jmmMetadata", gson.toJson(jmmMetadata))
            )
        }

        suspend fun NativeMicroModule.nativeFetchInstallApp(jmmMetadata: JmmMetadata, url: String) {
            nativeFetch(
                Uri.of("file://jmm.sys.dweb/install")
                    .query("mmid", jmmMetadata.id).query("metadataUrl", url)
            )
        }
    }

    init {
        // TODO 启动的时候，从数据库中恢复 apps 对象
        GlobalScope.launch(ioAsyncExceptionHandler) {
            JmmMetadataDB.queryJmmMetadataList().collectLatest {
                apps.clear()
                it.forEach { (key, value) ->
                    apps[key] = JsMicroModule(value)
                    nativeFetch(
                        Uri.of("file://dns.sys.dweb/install")
                            .query("jmmMetadata", gson.toJson(value))
                    )
                }
            }
        }
    }

    val queryMetadataUrl = Query.string().required("metadataUrl")
    val queryMmid = Query.string().required("mmid")

    override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {

        apiRouting = routes(
            "/install" bind Method.GET to defineHandler { request ->
                val metadataUrl = queryMetadataUrl(request)
                val jmmMetadata =
                    nativeFetch(metadataUrl).json<JmmMetadata>(JmmMetadata::class.java)
                // 根据 jmmMetadata 打开一个应用信息的界面，用户阅读界面信息后，可以点击"安装"
                openJmmMetadataInstallPage(jmmMetadata) /*{ metadata ->
                    JsMicroModule(metadata).apply {
                        apps[jmmMetadata.id] = this // 添加应用
                        bootstrapContext.dns.install(this) // 注册应用
                    }
                }*/
                return@defineHandler jmmMetadata
            },
            "/uninstall" bind Method.GET to defineHandler { request, ipc ->
                val mmid = queryMmid(request)
                val jmm = apps[mmid] ?: throw Exception("")
                openJmmMatadataUninstallPage(jmm.metadata)
                return@defineHandler true
            },
            "/query" bind Method.GET to defineHandler { request, ipc ->
                return@defineHandler AppsQueryResult(
                    apps.map { it.value.metadata },
                    installingApps.map { it.value })
            }
        )

    }

    data class AppsQueryResult(
        val installedAppList: List<JmmMetadata>,
        val installingAppList: List<InstallingAppInfo>
    )

    data class InstallingAppInfo(var progress: Float, val jmmMetadata: JmmMetadata)

    private val installingApps = mutableMapOf<Mmid, InstallingAppInfo>()
    private suspend fun openJmmMetadataInstallPage(
        jmmMetadata: JmmMetadata/*, installDNS: (JmmMetadata) -> Unit*/
    ) {
        // 打开安装的界面
        JmmManagerActivity.startActivity(jmmMetadata)
        // 拿到安装的状态
        /*val observe = DownLoadObserver(jmmMetadata.id)
        GlobalScope.launch(Dispatchers.IO) {
            observe.observe {
                if (it.downLoadStatus == DownLoadStatus.INSTALLED) {
                    installDNS(jmmMetadata)
                }
            }
        }*/
    }

    private fun openJmmMatadataUninstallPage(jmmMetadata: JmmMetadata) {
        TODO("Not yet implemented")
    }

    override suspend fun _shutdown() {

    }

}

File: ios/BFExplorer/Plaoc/IPCCore/sys/jmm/JmmNMM.swift
//
//  JmmNMM.swift
//  BFExplorer
//
//  Created by ui03 on 2023/3/2.
//

import Foundation
import Vapor
import HandyJSON

class JmmNMM: NativeMicroModule {
    
    static var apps: [String: JsMicroModule] = [:]
    private var installingApps: [String: InstallingAppInfo] = [:]
    
    static let hostName = "file://jmm.sys.dweb"
    static var jmmNMM: JmmNMM?
    
    init() {
        super.init(mmid: "jmm.sys.dweb")
        
        // TODO 启动的时候，从数据库中恢复 apps 对象
        Task {
            
        }
        
        JmmNMM.jmmNMM = self
    }
    
    static func getAndUpdateJmmNmmApps() -> [String: JsMicroModule] {
        return apps
    }
    
    static func nativeFetchFromJS(mmid: String) {
        Task {
            let response = self.apps[mmid]?.nativeFetch(urlstring: "file://dns.sys.dweb/open?app_id=\(mmid.encodeURIComponent())")
            if response == nil {
                print("JmmNMM no found jmm mmid \(mmid)")
            }
        }
    }
    
    static func nativeFetchOpenInstall(jmmMetadata: JmmMetadata, url: String) {
        Task {
            jmmNMM?.nativeFetch(urlstring: "\(hostName)/open?mmid=\(jmmMetadata.id)&metadataUrl=\(url)")
        }
    }
    
    static func nativeFetchInstallDNS(jmmMetadata: JmmMetadata) {
        Task {
            jmmNMM?.nativeFetch(urlstring: "\(hostName)/install?metadata=\(jmmMetadata.toJSONString() ?? "")")
        }
    }
    
    override func _bootstrap(bootstrapContext: BootstrapContext) throws {
        
        for app in JmmNMM.apps.values {
            bootstrapContext.dns.install(mm: app)
        }
        routerHandler(bootstrapContext: bootstrapContext)
    }
    
    private func routerHandler(bootstrapContext: BootstrapContext) {
        let installRouteHandler: RouterHandler = { request, ipc in
            
            let metadata = request.query[String.self, at: "metadata"] ?? ""
            guard let jmmMetadata = JSONDeserializer<JmmMetadata>.deserializeFrom(json: metadata) else {
                return false
            }
            
            var module = JmmNMM.apps[jmmMetadata.id]
            if module == nil {
                module = JsMicroModule(metadata: jmmMetadata)
                bootstrapContext.dns.install(mm: module!)
                JmmNMM.apps[jmmMetadata.id] = module
            }
            return true
        }
        let uninstallRouteHandler: RouterHandler = { request, ipc in
            let mmid = request.query[String.self, at: "mmid"] ?? ""
            let jmm = JmmNMM.apps[mmid]
            if jmm == nil {
                fatalError("")
            }
            self.openJmmMatadataUninstallPage(jmmMetadata: jmm!.metadata)
            return true
        }
        
        let queryRouteHandler: RouterHandler = { request, ipc in
            return AppsQueryResult(installedAppList: JmmNMM.apps.map { $1.metadata }, installingAppList: self.installingApps.map { $1 })
        }
        
        let openRouteHandler: RouterHandler = { request, ipc in
            let metadataUrl = request.query[String.self, at: "metadataUrl"] ?? ""
            let jmmMetadata = JSONDeserializer<JmmMetadata>.deserializeFrom(json: self.nativeFetch(urlstring: metadataUrl)?.body.string ?? "")
            self.openJmmMatadataInstallPage(jmmMetadata: jmmMetadata)
            return jmmMetadata
        }
        apiRouting["\(self.mmid)/install"] = installRouteHandler
        apiRouting["\(self.mmid)/uninstall"] = uninstallRouteHandler
        apiRouting["\(self.mmid)/query"] = queryRouteHandler
        apiRouting["\(self.mmid)/open"] = openRouteHandler
        
        // 添加路由处理方法到http路由中
        let app = HttpServer.app
        let group = app.grouped("\(mmid)")
        let httpHandler: (Request) throws -> Response = { request in
            self.defineHandler(request: request)
        }
        for pathComponent in ["install", "uninstall", "query", "open"] {
            group.on(.GET, [PathComponent(stringLiteral: pathComponent)], use: httpHandler)
        }
    }
    
    private func openJmmMatadataInstallPage(jmmMetadata: JmmMetadata?) {
        //TODO
    }
    
    private func openJmmMatadataUninstallPage(jmmMetadata: JmmMetadata?) {
       // TODO("Not yet implemented")
    }
    
    override func _shutdown() throws {
        
    }
}


struct InstallingAppInfo {
    
    var progress: Float
    var jmmMetadata: JmmMetadata
    
    init(progress: Float, jmmMetadata: JmmMetadata) {
        self.progress = progress
        self.jmmMetadata = jmmMetadata
    }
}


struct AppsQueryResult {
    
    var installedAppList: [JmmMetadata]
    var installingAppList: [InstallingAppInfo]
    
    init(installedAppList: [JmmMetadata], installingAppList: [InstallingAppInfo]) {
        self.installedAppList = installedAppList
        self.installingAppList = installingAppList
    }
}
